lib/guardian.ml

```diff
@@ -246,35 +222,28 @@ module Make (A : RoleSig) (T : RoleSig) = struct
       ;;
 
       let find_authorizable ?ctx (typ : 'kind) (id : Uuid.Actor.t) =
-        let open Lwt_result.Syntax in
-        let* mem = Authorizable.mem ?ctx id in
-        if mem
-        then
-          let* roles = find_roles ?ctx id in
-          let* owner = find_owner ?ctx id in
-          Lwt.return (Ok (AuthorizableActor.make ?owner roles typ id))
+        let open Lwt_result.Infix in
+        Authorizable.mem ?ctx id
+        >>= fun exists ->
+        if exists
+        then find ?ctx typ id
         else
-          Lwt.return
-            (Error
-               (Format.asprintf
-                  "Authorizable %s doesn't exist."
-                  (Uuid.Actor.to_string id)))
+          Lwt_result.fail
+            (Format.asprintf
+               "Authorizable %s doesn't exist."
+               (Uuid.Actor.to_string id))
       ;;
 
       (** [save_rules rules] adds all the permissions [rules] to the backend. If
           there is an error at any point, it returns a `result` containing all
           of the items that were not added. *)
-      let save_rules ?ctx rules =
-        CCList.fold_left
+      let save_rules ?ctx =
+        Lwt_list.fold_left_s
           (fun acc x ->
-            match%lwt acc with
-            | Ok acc' ->
-              (match%lwt save_rule ?ctx x with
-               | Ok () -> Lwt.return_ok (x :: acc')
-               | Error _ -> Lwt.return_error [ x ])
-            | Error xs -> Lwt.return_error (x :: xs))
-          (Lwt.return_ok [])
-          rules
+            match%lwt save_rule ?ctx x with
```

Comment:
I think this changes the semantics slightly. In your version `save_rule` is called for every element, whereas in the old version once the first error is encountered for `save_rule`, `acc` will always be `Error`, so `save_rule` is never run again. This would be fine if everything were pure here, but we are doing `Lwt`.

Reply:
Ah right, the new version is more true to the docstring, let's keep that